mars/dataframe/indexing/getitem.py

```diff
@@ -194,28 +198,90 @@ class DataFrameIndex(DataFrameOperand, DataFrameOperandMixin):
     _op_type_ = OperandDef.INDEX
 
     _col_names = AnyField('col_names')
+    _mask = AnyField('mask')
 
-    def __init__(self, col_names=None, object_type=ObjectType.series, **kw):
-        super(DataFrameIndex, self).__init__(_col_names=col_names, _object_type=object_type, **kw)
+    def __init__(self, col_names=None, mask=None, object_type=ObjectType.series, **kw):
+        super(DataFrameIndex, self).__init__(_col_names=col_names, _mask=mask,
+                                             _object_type=object_type, **kw)
 
     @property
     def col_names(self):
         return self._col_names
 
+    @property
+    def mask(self):
+        return self._mask
+
     def __call__(self, df):
-        # if col_names is a list, return a DataFrame, else return a Series
-        if isinstance(self._col_names, list):
-            dtypes = df.dtypes[self._col_names]
-            columns = parse_index(pd.Index(self._col_names), store_data=True)
-            return self.new_dataframe([df], shape=(df.shape[0], len(self._col_names)), dtypes=dtypes,
-                                      index_value=df.index_value, columns_value=columns)
+        if self.col_names is not None:
+            # if col_names is a list, return a DataFrame, else return a Series
+            if isinstance(self._col_names, list):
+                dtypes = df.dtypes[self._col_names]
+                columns = parse_index(pd.Index(self._col_names), store_data=True)
+                return self.new_dataframe([df], shape=(df.shape[0], len(self._col_names)), dtypes=dtypes,
+                                        index_value=df.index_value, columns_value=columns)
+            else:
+                dtype = df.dtypes[self._col_names]
+                return self.new_series([df], shape=(df.shape[0],), dtype=dtype, index_value=df.index_value,
+                                    name=self._col_names)
         else:
-            dtype = df.dtypes[self._col_names]
-            return self.new_series([df], shape=(df.shape[0],), dtype=dtype, index_value=df.index_value,
-                                   name=self._col_names)
+            if isinstance(self.mask, SERIES_TYPE):
+                return self.new_dataframe([df, self._mask], shape=(np.nan, df.shape[1]), dtypes=df.dtypes,
+                                          index_value=parse_index(pd.Index([])), columns_value=df.columns)
```

Comment:
Seems that the `index_value` should keep the df's one? And some tests may be added to test_indexing to verify the correctness of expr part.

Reply:
The `index_value` is normally different with df's one. For example:

```python
>>> df = pd.DataFrame({'x': [1,2,3]})
>>> ss = pd.Series([True, False, True])
>>> df.index
RangeIndex(start=0, stop=3, step=1)
>>> df[ss].index
Int64Index([0, 2], dtype='int64')
```

Reply:
Yeah, for `RangeIndex`, it will be Int64Index, but if `pd.Index` the dtype will be lost, right? And I have some concern that

```
In [65]: from mars.dataframe.utils import parse_index                           

In [67]: parse_index(pd.Index([])).key                                          
Out[67]: '8050bd4a54d862bbcdd7bb78e6c08fc1'

In [68]: parse_index(pd.Index([])).key                                          
Out[68]: '8050bd4a54d862bbcdd7bb78e6c08fc1'
```

The key is the same, maybe we should generate new key according to mask's key?

Reply:
Fixed, and test case has been added.